flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/BaseConversionUtils.java

```diff
@@ -0,0 +1,175 @@
+        long val = 0;
+        long bound = unsignedLongDiv(-1 - radix, radix); // Possible overflow once
+        // val
+        // exceeds this value
+        for (int i = fromPos; i < value.length && value[i] >= 0; i++) {
+            if (val >= bound) {
+                // Check for overflow
+                if (unsignedLongDiv(-1 - value[i], radix) < val) {
+                    return -1;
+                }
+            }
+            val = val * radix + value[i];
+        }
+        return val;
+    }
+
+    /**
+     * Convert the bytes in value[] to the corresponding chars.
+     *
+     * @param radix must be between MIN_RADIX and MAX_RADIX
+     * @param fromPos is the first nonzero element
+     */
+    private static void byte2char(byte[] value, int radix, int fromPos) {
+        for (int i = fromPos; i < value.length; i++) {
+            value[i] = (byte) Character.toUpperCase(Character.forDigit(value[i], radix));
+        }
+    }
+
+    /**
+     * Convert the chars in value[] to the corresponding integers. Convert invalid characters to -1.
+     *
+     * @param radix must be between MIN_RADIX and MAX_RADIX
+     * @param fromPos is the first nonzero element
+     */
+    private static boolean char2byte(byte[] value, int radix, int fromPos) {
+        for (int i = fromPos; i < value.length; i++) {
+            value[i] = (byte) Character.digit(value[i], radix);
+            if (value[i] == -1) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Convert numbers between different number bases. If toBase&gt;0 the result is unsigned,
+     * otherwise it is signed.
+     */
+    public static byte[] conv(byte[] n, int fromBase, int toBase) {
+        byte[] value = new byte[64];
+        if (n == null || n.length < 1) {
+            return null;
+        }
+
+        if (fromBase < Character.MIN_RADIX
+                || fromBase > Character.MAX_RADIX
+                || Math.abs(toBase) < Character.MIN_RADIX
+                || Math.abs(toBase) > Character.MAX_RADIX) {
+            return null;
+        }
+
+        boolean negative = (n[0] == '-');
+        int first = 0;
+        if (negative) {
+            first = 1;
+        }
+
+        // Copy the digits in the right side of the array
+        for (int i = 1; i <= n.length - first; i++) {
+            value[value.length - i] = n[n.length - i];
+        }
+        if (!char2byte(value, fromBase, value.length - n.length + first)) {
+            return null;
+        }
+
+        // Do the conversion by going through a 64 bit integer
+        long val = encode(value, fromBase, value.length - n.length + first);
+        if (negative && toBase > 0) {
+            if (val < 0) {
+                val = -1;
+            } else {
+                val = -val;
+            }
+        }
+        if (toBase < 0 && val < 0) {
+            val = -val;
+            negative = true;
+        }
```

Comment:
Hello, I'm currently working on this function. I find the logic of the code from lines 151 to 161 very strange, which is the same as hive and spark, do you have any insights on it?